src/repl/scala/tools/nsc/interpreter/Naming.scala

```diff
@@ -12,52 +12,56 @@
 
 package scala.tools.nsc.interpreter
 
-import scala.util.Properties.lineSeparator
 import scala.util.matching.Regex
 
 /** This is for name logic which is independent of the compiler (notice there's no Global.)
  *  That includes at least generating, metaquoting, mangling, and unmangling.
  */
 object Naming {
-  def unmangle(str: String): String = {
-    val ESC = '\u001b'
-    val cleaned = lineRegex.replaceAllIn(str, "")
-    // Looking to exclude binary data which hoses the terminal, but
-    // let through the subset of it we need, like whitespace and also
-    // <ESC> for ansi codes.
-    val binaryChars = cleaned count (ch => ch < 32 && !ch.isWhitespace && ch != ESC)
-    // Lots of binary chars - translate all supposed whitespace into spaces
-    // except supposed line endings, otherwise scrubbed lines run together
-    if (binaryChars > 5) // more than one can count while holding a hamburger
-      cleaned map {
-        case c if lineSeparator contains c => c
-        case c if c.isWhitespace => ' '
-        case c if c < 32 => '?'
-        case c => c
-      }
-    // Not lots - preserve whitespace and ESC
-    else
-      cleaned map (ch => if (ch.isWhitespace || ch == ESC) ch else if (ch < 32) '?' else ch)
-  }
+//For CSI, or "Control Sequence Introducer" commands, the ESC [ is followed by any number (including none) of "parameter bytes" in the range 0x30–0x3F (ASCII 0–9:;<=>?), then by any number of "intermediate bytes" in the range 0x20–0x2F (ASCII space and !"#$%&'()*+,-./), then finally by a single "final byte" in the range 0x40–0x7E (ASCII @A–Z[\]^_`a–z{|}~)
+  private final val esc = "\u001b"    // "\N{escape}"
+  private val csi = raw"$esc\[[0-9;]*([\x40-\x7E])"
+  private lazy val cleaner = raw"$csi|([^\p{Print}\p{Space}]+)|$linePattern".r
```

Comment:
What's the second pattern for?  1-n non-print/space characters?

Reply:
Yes, just trying for a reasonable test. The existing ancient heuristic is too brittle.